### PR TITLE
[ENT-341] Change `enterprise-catalogs` & `programs` endpoints to meet requirements

### PR DESCRIPTION
**Pull request by UmanShahzad**
**Description:** This PR modifies the initial work done for ENT-341 in the following ways:
1. Added the EnterpriseCustomerCatalog.title field.
2. Replaced the EnterpriseCustomerCatalog.query field with the EnterpriseCustomerCatalog.content_filter field. The content_filter field is a JSONField which stores query parameters that will be used to filter the search/all request made to the discovery service.
3. Removed the /enterprise/api/v1/programs/{uuid}/ endpoint.
4. Added the following endpoints:
/enterprise/api/v1/enterprise-catalogs/{uuid}/course-runs/{uuid}/
/enterprise/api/v1/enterprise-catalogs/{uuid}/programs/{uuid}/
These endpoints return the full metadata associated with the given resources instead of the condensed version returned from the search/all endpoint. Also, these endpoints will make sure the resource being requested is actually a member of the EnterpriseCustomerCatalog included in the request.
5. Modified the /enterprise/api/v1/enterprise-catalogs/{uuid}/ endpoint to return the EntepriseCustomerCatalog fields along with the paginated results of the request made to the search/all endpoint on the discovery service.
6. All requests made to any of the enterprise-catalogs endpoints will ensure that the requesting user is linked to the EnterpriseCustomerCatalog via the EnterpriseCustomerCatalog -> EnterpriseCustomer -> EnterpriseCustomerUser data model relationship.

**JIRA:** [ENT-341](https://openedx.atlassian.net/browse/ENT-341)

**Dependencies:** None.

**Merge deadline:** Asap.

**Testing instructions:**

1. Identify/create both a staff and non-staff user which you will use for testing the new endpoints.
2. [Link](https://business.sandbox.edx.org/admin/enterprise/enterprisecustomeruser/add/) both your staff and non-staff users to the PiedPiper EnterpriseCustomer.
3. [Identify/create](https://business.sandbox.edx.org/admin/enterprise/enterprisecustomercatalog/) EnterpriseCustomerCatalog models linked and unlinked with the PiedPiper EnterpriseCustomer.
4. Login as staff and access the https://business.sandbox.edx.org/enterprise/api/v1/enterprise-catalogs/ endpoint. Verify that all EnterpriseCustomerCatalogs are returned in the result.
5. Access the https://business.sandbox.edx.org/enterprise/api/v1/enterprise-catalogs/{uuid} endpoint replacing the {uuid} with one of the ECC UUIDs returned from the enterprise-catalogs endpoint. Verify that the appropriate results are returned based on the content_filter configured for the EnterpriseCustomerCatalog used.
6. Access the https://business.sandbox.edx.org/enterprise/api/v1/enterprise-catalogs/{ecc_uuid}/programs/{program_uuid} endpoint replacing UUIDs with appropriate UUIDs returned from the previous API calls. Verify that program detail is returned.
7. Access the https://business.sandbox.edx.org/enterprise/api/v1/enterprise-catalogs/{ecc_uuid}/course-runs/{course_run_id} endpoint replacing IDs with appropriate IDs returned from the previous API calls. Verify that course run detail is returned.
8. Repeat steps 4-7 with a non-staff user that is linked to the PiedPiper EnterpriseCustomer.
9. Repeat steps 4-7 with a non-staff user that is not-linked to the PiedPiper EnterpriseCustomer. Verify that results are NOT returned.

**Merge checklist:**

- [ ] Check that the versions of the requirements in the `platform-master.in` file match edx-platform.
- [ ] New requirements are in the right place (`base.in` if only used in enterprise; in the correct `platform-****.in` files if they're hosted in edx-platform)
- [ ] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are (reasonably) squashed
- [ ] Translations are updated
- [ ] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)
- [ ] edx-platform PR (be sure to include edx-platform requirements upgrades that were present in this PR)

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @UmanShahzad! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-enterprise&number=170"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

**Comment by douglashall** [comment]
@brittneyexline I haven't given this a full review yet, but it looks like a good starting point for the remaining enterprise programs API work. If you want to make progress on this while I am out, you could add code that decorates the program data with the enrollment URL which will navigate the learner to the program enrollment landing page.

**Comment by UmanShahzad** [comment]
@douglashall @brittneyexline I'm going to soon be manually testing this as it is through #171 because I'll be needing it for the programs landing page, FYI. Will post my results on here and there for completeness.

**Comment by UmanShahzad** [comment]
@douglashall @brittneyexline I gave this a test and the endpoints work as we expect, all 4 of

- `enterprise-catalogs-list`
- `enterprise-catalogs-detail`
- `programs-list`
- `programs-detail`

We can put this on the sandbox to get further quality assurance, but LGTM!

**Comment by UmanShahzad** [comment]
Squashed from 326cfde75dd909d6aee5028126721baa2fc3f5ec

**Comment by UmanShahzad** [comment]
@brittneyexline @douglashall @zrana This is now on the sandbox.

**Comment by UmanShahzad** [comment]
I tried this on the sandbox and all endpoints work but, for `/enterprise/api/v1/enterprise-catalogs/`, it's not showing all the results. This is due to the filtering I do here:

```python
        self.queryset = self.queryset.filter(
            enterprise_customer__enterprise_customer_users__user_id=request.user.id
        )
```

The staff user I used on the sandbox isn't linked to an enterprise customer that any of the existing `EnterpriseCustomerCatalog`s are linked to, but if so, it works.

I'm not sure if this is the right filter to use, @douglashall? I think it is, but just making sure.

Also, we already check for permissions, i.e. is the user a staff member or whether there exists a catalog that the requesting user is linked to, but I just realized our permission check is loose: it checks if there's *any* catalog that the requesting user is linked to! It then would allow the user to make a specific query `/enterprise/api/v1/enterprise-catalogs/uuid/` and retrieve the search result of any other enterprise catalog. So we need to tighten this constraint, which is editing a single line, but before I do that, I just want to be sure this is the correct constraint, as well as the above, @douglashall.

**Comment by UmanShahzad** [comment]
@douglashall I actually just started working on it anyway, and the solution is in the latest commit, which is NOT on the sandbox yet.

**Comment by douglashall** [comment]
@UmanShahzad Thanks for outlining the permissions issues here. It sounds like you have the right approach for addressing them. Let me know when you are done working on this branch for the day. I should have bandwidth for review/completing remaining issues today.

**Comment by UmanShahzad** [comment]
@douglashall I am done with it for the day, yes :) Thanks for taking this over!

**Comment by itsjeyd** [comment]
@douglashall Are you going to be making changes to the API gateway as part of the follow-up work you're doing here? And if so, will they require additional PRs against other repos (such as https://github.com/edx/api-manager)?

**Comment by douglashall** [comment]
@itsjeyd Yes and yes.

**Comment by itsjeyd** [comment]
@douglashall Alright, thanks for the quick reply :)

File: enterprise/models.py [hunk 1]
```diff
@@ -223,6 +223,28 @@ def get_course_enrollment_url(self, course_run_key):
...
+                _("This package must be installed in an EdX environment to look up configuration.")
+            )
+
+        return urljoin(
+            configuration_helpers.get_value('LMS_ROOT_URL', settings.LMS_ROOT_URL),
+            reverse(
+                'enterprise_program_enrollment_page',
+                kwargs={'enterprise_uuid': self.uuid, 'program_uuid': program_uuid}
+            )
+        )
```
Comment (UmanShahzad) [inline enterprise/models.py]: I like this, thanks for putting it!

**Comment by UmanShahzad** [comment]
@douglashall Just let me know when this is ready for review and I'll take a jump on it, including testing.

File: api-compact.yaml [hunk 2]
```diff
@@ -12,7 +12,8 @@
 #   /enterprise/api/v1/catalogs/{id}/courses/
 #   /enterprise/api/v1/enterprise-catalogs/
 #   /enterprise/api/v1/enterprise-catalogs/{uuid}/
-#   /enterprise/api/v1/programs/{uuid}/
+#   /enterprise/api/v1/enterprise-catalogs/{uuid}/course-runs/{uuid}/
```
Comment (UmanShahzad) [inline api-compact.yaml]: Should be `id` rather than `uuid` I believe.

File: enterprise/api_client/discovery.py [hunk 3]
```diff
@@ -284,6 +284,22 @@ def get_program_course_keys(self, program_uuid):
             return []
         return [course['key'] for course in program_details.get('courses', [])]
 
+    def get_paginated_programs(self, querystring=None):
```
Comment (UmanShahzad) [inline enterprise/api_client/discovery.py]: Is this now unused btw?

File: enterprise/models.py [hunk 4]
```diff
@@ -225,6 +227,30 @@ def get_course_enrollment_url(self, course_run_key):
...
+
+        return urljoin(
+            configuration_helpers.get_value('LMS_ROOT_URL', settings.LMS_ROOT_URL),
+            '/enterprise/{}/program/{}/enroll'.format(self.uuid, program_uuid)
+            # TODO: Replace above with the reverse statement below, once the program enrollment landing page is created.
+            # reverse(
+            #     'enterprise_program_enrollment_page',
+            #     kwargs={'enterprise_uuid': self.uuid, 'program_uuid': program_uuid}
+            # )
+        )
```
Comment (UmanShahzad) [inline enterprise/models.py]: Thanks for the TODO, I'll update that.

**Comment by UmanShahzad** [comment]
@douglashall I have yet to test this, but I have a question after taking a look at the resulting endpoints: how am I supposed to know which `EnterpriseCustomerCatalog` UUID to use when querying for `programs` on the program enrollment landing page? Previously we had the `/api/v1/programs/{uuid}/` endpoint that didn't require it, but now since a customer can have multiple catalogs, I'm not sure how to retrieve program information if I'm supposed to use `/api/v1/enterprise-catalogs/{uuid}/programs/{uuid}/`.

**Comment by douglashall** [comment]
@UmanShahzad When retrieving program metadata for the program enrollment landing page, we should go directly to the discovery service and use the program detail API endpoint there. No need to call the EnterpriseCustomerCatalog endpoint. The EnterpriseCustomerCatalog endpoints are for third-party consumption of an EnterpriseCustomer's catalog metadata.

**Comment by UmanShahzad** [comment]
@douglashall Ok, that makes sense. I was told before to be using our wrapper programs endpoint (`/enterprise/api/v1/programs/{uuid}/`), but that seems to be old news now with the changes.

**Comment by UmanShahzad** [comment]
@brittneyexline I know you +1'd this before, but it's changed significantly since last time, so if you get time to review this again that'd be great and we could merge it :)

File: api-compact.yaml (hunk 2 above)
Comment (douglashall) [reply api-compact.yaml]: Yes. I will update. Thanks.

File: enterprise/api_client/discovery.py (hunk 3 above)
Comment (douglashall) [reply enterprise/api_client/discovery.py]: Yes. I will remove this and associated test code. Thanks.

**Pull request merged**